factory-containers/docker-app

```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/python3
+import json
+import os
+import subprocess
+import sys
+
+import docker
+import yaml
+
+from typing import Iterable
+
+from helpers import cmd, require_env, status
+
+
+def convert_file(path: str):
+    """Take a .dockerapp file to directory format.
+
+       The file format isn't supported by docker-app 0.8 and beyond. Just
+       split a 3 sectioned yaml file into its pieces
+    """
+    with open(path) as f:
+        meta, compose, params = yaml.safe_load_all(f)
+
+    os.unlink(path)
+    os.mkdir(path)
+
+    with open(os.path.join(path, 'metadata.yml'), 'w') as f:
+        yaml.dump(meta, f)
+    with open(os.path.join(path, 'docker-compose.yml'), 'w') as f:
+        yaml.dump(compose, f)
+    with open(os.path.join(path, 'parameters.yml'), 'w') as f:
+        yaml.dump(params, f)
+
+
+def pin_versions(compose_file: str, factory: str, arch_tag: str):
+    """Pin all containers produced in this factory to the $TAG-$ARCH
+
+       docker-app actually tries to pin things. However, this script gets
+       called before we've push the multi-arch image our registry. So
+       we need to pin to something we know at the time this is run.
+    """
+    with open(compose_file) as f:
+        data = yaml.safe_load(f)
+
+    factory_container_base = 'hub.foundries.io/' + factory + '/'
+    for name, svc in data['services'].items():
+        img = svc['image']
+        if img.startswith(factory_container_base) and img.endswith(':latest'):
+            img = img.replace(':latest', ':' + arch_tag)
+            status('Pinning ' + img)
+            svc['image'] = img
+
+    with open(compose_file, 'w') as f:
+        yaml.dump(data, f)
+
+
+def iterate_apps(factory: str, container_tag: str) -> Iterable[str]:
+    """Loop through all the docker apps in a factory and pin the relevent
+       container images. If the its a <0.9 single file dockerapp, convert
+       in place to the new directory based format.
+    """
+    for p in os.listdir():
+        if not p.endswith('.dockerapp'):
+            continue
+        if os.path.isfile(p):
+            status('Converting .dockerapp file for: ' + p)
+            convert_file(p)
+
+        status('Pinning docker-compose images for: ' + p)
+        pin_versions(
+            os.path.join(p, 'docker-compose.yml'), factory, container_tag)
+
+        yield p
+
+
+def build_app(dockerapp_path: str, factory: str, tag: str) -> str:
```

Comment:
We need to check the docker-build.conf before attempting to build, in case we skip building.

Reply:
@EmbeddedAndroid - I think i have this fixed now. My test looks like this:

 https://ci.foundries.io/projects/andy-corp/lmp/builds/324/